src/cargo/core/dependency.rs

```diff
@@ -303,8 +303,20 @@ impl Dependency {
     }
 
     /// Lock this dependency to depending on the specified package id
-    pub fn lock_to(self, id: &PackageId) -> Dependency {
-        self.clone_inner().lock_to(id).into_dependency()
+    pub fn lock_to(mut self, id: &PackageId) -> Dependency {
+        Rc::make_mut(&mut self.inner).lock_to(id);
+        self
+    }
+
+    /// Set the version requirement for this dependency
+    pub fn set_version_req(&mut self, req: VersionReq) -> &mut Dependency {
+        Rc::make_mut(&mut self.inner).set_version_req(req);
+        self
+    }
+
+    pub fn set_kind(&mut self, kind: Kind) -> &mut Dependency {
```

Comment:
I don't really like that these two methods are duplicated between `Dependency` and `DependencyInner`. Looks like `DependencyInner` plays dual role of a dependency builder (it exposes more setters then `Dependency`) and an Rc wrappeed data store. Also, there are a bunch of `pub` gettrs on `DependencyInner` which can be private:

```Rust
    fn version_req(&self) -> &VersionReq { &self.req }
    fn name(&self) -> &str { &self.name }
    fn source_id(&self) -> &SourceId { &self.source_id }
    fn kind(&self) -> Kind { self.kind }
    fn specified_req(&self) -> bool { self.specified_req }
    fn platform(&self) -> Option<&Platform> {
        self.platform.as_ref()
    }
```

Ouch, and `lock_to` on `DependencyInner` can be private as well!

Could we somehow improve Dependency/DependencyInner separation? Perhaps we can make a private `DependencyData` without any getters as a thing inside `Rc` inside `Dependency`, and a public `DependencyBuilder` with all public fields and no getters, so that one can use sweet `{.. default}` syntax to construct deps without setters boilerplate. 

Conflating refactorings and perf optimizations is not necessary a good idea, but perhaps we can make things nicer in a simple way by just moving all those "can be private" things from `DependencyInner` directly into `Dependency`, just to offset the duplication in `set_version` and `set_kind`?